### PR TITLE
fix: add null check for connection.headers

### LINKED ISSUE #2199: TypeError: Cannot read property 'headers' of null at Server.socket.on (***/node_modules/webpack-dev-server/lib/servers/SockJSServer.js:68:32)
- Operating System: Windows 10
- Node Version: 12.8.1
- NPM Version: 6.10.3
- webpack Version: 4.39.2
- webpack-dev-server Version: since 3.5.0
- Browser: Chrome 76

- [X] This is a **bug**
- [ ] This is a **modification** request

### Expected Behavior

`connection` should not be null

### Actual Behavior

`TypeError: Cannot read property 'headers' of null at Server.socket.on (***/node_modules/webpack-dev-server/lib/servers/SockJSServer.js:68:32)`

### For Bugs; How can we reproduce the behavior?

This bug is difficult to reproduce. It happens occasionally.

https://github.com/webpack/webpack-dev-server/blob/2d44ef8088db2312516274cc2d5edda348064712/lib/servers/SockJSServer.js#L65-L69

line 67 should become

```js
if (connection) {
  f(connection, connection.headers)
}
```

### PR DESCRIPTION
**Pull request by wood1986**
fix #2199

- [X] This is a **bugfix**
- [ ] This is a **feature**
- [ ] This is a **code refactor**
- [x] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### CODE REVIEW

**Comment by Loonride** [comment]
If the connection object is null, there is no use in passing it along to the callback since we can't do anything with it. I think it is better off being:

```
if (connection) {
  f(connection, connection.headers);
}
```

**Comment by wood1986** [comment]
> If the connection object is null, there is no use in passing it along to the callback since we can't do anything with it. I think it is better off being:
> 
> ```
> if (connection) {
>   f(connection, connection.headers);
> }
> ```

I thought about this. As you have a check in

https://github.com/webpack/webpack-dev-server/blob/a3558944b900f5f8695421522eeb102711b81ce7/lib/Server.js#L662-L672

I want to reuse.

**Comment by wood1986** [comment]
> need tests

I added

**Comment by wood1986** [comment]
I need some help to re-queue the build

**Comment by wood1986** [comment]
Any updates?

File: test/server/servers/SockJSServer.test.js [hunk 1]
```diff
@@ -84,9 +84,20 @@ describe('SockJSServer', () => {
...
       }, 3000);
     });
+
+    it('should not throw an exception when connection is null', (done) => {
+      expect(() => {
+        socketServer.onConnection(() => {});
+        socketServer.socket.emit('connection', null);
+      }).not.toThrow();
+
+      setTimeout(() => {
```
Comment (Loonride) [inline test/server/servers/SockJSServer.test.js]: This timeout is not needed, remove this and remove the `done` parameter from the function.

File: test/server/servers/SockJSServer.test.js [hunk 2]
```diff
@@ -84,9 +84,20 @@ describe('SockJSServer', () => {
         done();
       }, 3000);
     });
+
+    it('should not throw an exception when connection is null', (done) => {
+      expect(() => {
+        socketServer.onConnection(() => {});
```
Comment (Loonride) [inline test/server/servers/SockJSServer.test.js]: Before this `expect` test, could you make a function `const cb = jest.fn()`, then pass it into the onConnection like: `onConnection(cb)`. Then after this confirm that the callback is actually being called with the expected values like: `expect(cb.mock.calls[0]).toEqual([null, null])`.

File: test/server/servers/SockJSServer.test.js [hunk 3]
```diff
@@ -10,7 +10,7 @@ describe('SockJSServer', () => {
   let socketServer;
   let listeningApp;
 
-  beforeAll((done) => {
+  beforeEach((done) => {
```
Comment (Loonride) [inline test/server/servers/SockJSServer.test.js]: These can stay as `beforeAll` and `afterAll`. It is reusing the same server for all tests. Switching to `beforeEach/afterEach` would not change anything because there is only that one `describe` call on the same level as these calls.

**Comment by Loonride** [comment]
> I want to reuse.

Understood. I have added some notes on the tests.

File: test/server/servers/SockJSServer.test.js (hunk 3 above)
Comment (wood1986) [reply test/server/servers/SockJSServer.test.js]: * Reason for using `beforeEach`:
As my test needs to test null scenario and we are mutating the same environment, it shows
image

* Reason for the `timeout`
It won't pass in the build machine only for Windows node-8 and Windows node-6.
image
See https://dev.azure.com/webpack/webpack-dev-server/_build/results?buildId=5531

File: test/server/servers/SockJSServer.test.js (hunk 1 above)
Comment (wood1986) [reply test/server/servers/SockJSServer.test.js]: * Reason for the `timeout`
It won't pass in the build machine only for Windows node-8 and Windows node-6.
image
See https://dev.azure.com/webpack/webpack-dev-server/_build/results?buildId=5531

File: test/server/servers/SockJSServer.test.js (hunk 3 above)
Comment (Loonride) [reply test/server/servers/SockJSServer.test.js]: Yes that makes sense. My understanding of `beforeEach/afterEach` in describe blocks was incorrect.

File: test/server/servers/SockJSServer.test.js (hunk 1 above)
Comment (Loonride) [reply test/server/servers/SockJSServer.test.js]: That is caused by a stability issue with our e2e tests. You can ignore it, your changes are not causing it (you can see that your test ran and passed after the problematic test), and your changes can still be merged if CI is failing for this reason. The `timeout` can still be removed.

Comment (wood1986) [reply test/server/servers/SockJSServer.test.js]: Thanks. I am going to make the change

Comment (wood1986) [reply test/server/servers/SockJSServer.test.js]: Yes, it was the stability issue. Right now, all are passed. Thanks

File: test/server/servers/SockJSServer.test.js [hunk 4]
```diff
@@ -84,9 +84,18 @@ describe('SockJSServer', () => {
         done();
       }, 3000);
     });
+
+    it('should not throw an exception when connection is null', () => {
+      expect(() => {
+        const cb = jest.fn();
```
Comment (Loonride) [inline test/server/servers/SockJSServer.test.js]: Sorry one more nitpick but could you move this `cb` definition above the `expect` line, then move `expect(cb.mock.calls[0]).toEqual([null, null]);` below the `not.toThrow`? Just to isolate the behavior we want to "not throw".

Comment (wood1986) [reply test/server/servers/SockJSServer.test.js]: nvm and I love it

**Comment by Loonride** [comment]
Looks good to me /cc @evilebottnawi @hiroppy

**Comment by wood1986** [comment]
When will you guys do the npm publish?
